### PR TITLE
RealisticCaloDigi: new ROC integration

### LINKED ISSUE #41: Time realism of Calo Digi
The current implementation in the RealisticCaloDigi is estimated the time of a digitized hit as the time of the MC contribution with the earliest time:
https://github.com/iLCSoft/MarlinReco/blob/master/CaloDigi/Realistic/src/RealisticCaloDigi.cc#L291

It was pointed out that a more realistic time calculation should be estimated as follows:
- Sort MC contributions by increasing time
- Loop over contributions and accumulate deposit energy until you reach the digi threshold
- Estimate the time of the digitized hit as the time of the contribution that first reach the threshold

Does this have to be changed, or are we OK with the current (approximate) implementation ?

### PR DESCRIPTION
**Pull request by rete**
BEGINRELEASENOTES
- RealisticCaloDigi: Added new option for energy/charge and time integration
  - deal with slow/fast shaper of ROC chips and time estimate
  - new processor parameters:
     - **integrationMethod**: "Standard" for old implementation (default) or "ROC" emulating the behavior of ROC chip.
     - **fastShaper**: fast shaper time, unit in ns. Only for ROC method
     - **slowShaper**: slow shaper time, unit in ns. Only for ROC method
     - **timingResolution**: optional time resolution gaussian smearing to apply, unit . Only apply is > 0. Default is 0 (no smearing)

ENDRELEASENOTES

Fixes #41 

Parameter values for ROC implementation:
- AHCAL
   - **fastShaper** = 15 ns
   - **slowShaper** =  50 ns
   - **timingResolution** = 700 ps
- SiWECal
   - **fastShaper** = 90 ns (first estimate)
   - **slowShaper** =  180 ns (first estimate)
   - **timingResolution** = 700 ps (first guess from AHCAL value and the fact that the TDC is similar. Not a real estimate)

TODO:
- [x] Add details in release notes
- [x] Validate changes in the code for previous version (identical distributions)
- [x] Determine whether the ROC implementation works for the SiWECal electronics
- [x] Determine parameters for all technologies
- [x] Make plots:
    - rec time distribution
    - rec energy distribution
- [x] Document timing treatment in a Readme.md

### CODE REVIEW

**Comment by rete** [comment]
@gaede @ebrianne @dudarboh @JennyListDESY @danieljeans 
I would like to start a discussion on the implementation of a realistic time estimation in the ILD digitizers. Here is a PR, flagged as WIP. I will update the release notes to give more details soon. The original implementation of the ROC integration was taken from there: https://stash.desy.de/projects/CALICE/repos/calice_sim/browse/digitization/digiAhc2/src/Ahc2ROCThresholdProcessor.cc
Note that with the default settings at the moment, the old behaviour is kept for backward compatibility.

**Comment by danieljeans** [comment]
this must be done in very close collaboration with the hardware groups, and should explicitly be signed off by them.
I think it's probably a non-trivial thing to get right.

**Comment by ebrianne** [comment]
I have been already discussing with @rete about it yesterday and explaining why and how it is done for the AHCAL in TB. 
I agree that this would need to be validated by each hardware groups before the switch. I think we did studies some time ago on the impact of the threshold on timing which is minimal but this will I guess depend on the time resolution, especially at the ps level and may need to be looked into again. 
However, one has to be careful, we noticed not so long ago that the threshold has quite some impact on the energy below 1 MIP (at the 0.5 MIP threshold border). This is due to how we do our digitization workflow. You can contact Olin about this as he did the study and see if this would impact the ILDDigi workflow.

**Comment by rete** [comment]
@airqui ping

**Comment by airqui** [comment]
Hi @rete , thanks for the ping. I am indeed interested on this, specially on the test beam context, but I always manage to push it down in the pipeline of things to do...
I am up for discussions on this and to contribute to the si-ECAL part...

**Comment by rete** [comment]
The new implementation (ROC) works for the AHCAL electronics, as most of the code was taken from the AHCAL TB digitizer.
@airqui Would you mind going through the implementation and see whether it would work for the ECal electronics too?
Also we would need to determine a few parameter values as quoted in the top-level PR message. I assumed from a discussion with @ebrianne that the HCal values are correct.

**Comment by airqui** [comment]
Hi @rete,
sorry for the late reply. In the SiECAL, the efforts on the simulation/tb analysis have been somehow limited in the latest times since the personpower has been focused on the hardware activities. I plan to start looking at the digitization this January and hopefully we could conduct some comparison with old data during the first months of the year. This means that:

1. In January/February I will carefully check how we can adapt the AHCAL TB digitizer to our purposes (in principle our digitize is a simplified version of theirs)
2. The optimization of the needed parameters will take some time and I am not in position of making an educated guess of how long will it take..

Maybe it is not needed that we fully optimize the TB digitizer before work in the ILD digitizer. 

Question for @ebrianne : I imagine that the best way to start (thinking on test beam data) is to start with your AHCAL TB digitizer, right? Better than going directly to the ILD digitizer... am I correct?

**Comment by ebrianne** [comment]
I think this is the easiest way to start things. Technically I guess not much needs to be adapted but this depends on your framework I guess.

**Comment by rete** [comment]
@ebrianne @airqui @danieljeans  I have checked that the old integration (standard) is backward compatible. It looks all fine. 

In the new ROC implementation there seems to be a feature that I have not taken into account. As we now integrate using slow/fast shapers time parameters, there is not cut on the absolute hit time anymore. You can see it below for 30 GeV K0L, hcal hits only.

Full plot:

Plot_MyRecHCalLCTuple_Time

Zoom around the 100 ns cut value:

Plot_MyRecHCalLCTuple_TimeZoom

I have thus added in the latest commits (https://github.com/iLCSoft/MarlinReco/pull/83/commits/918489f81ae184639d3ec742a3dfd198163f2b5a) a similar time window cut using the same processor parameters for the ROC implementation. But now I'm not sure about the meaning of this cut in terms of electronics behavior. These late hits have to be cut away in any case but does this cut, **in the way it is implemented now**, make sense for the electronics. Isn't there a more meaningful way of applying this cut ?

As we've been discussing internally at DESY also, I will merge this pull request now, as time is getting short for me and as we will be running the standard integration in any case in the production and for some time. 

Give me your comments ! :-)

**Comment by gaede** [comment]
Hi Remi, as far as I understand it, this looks good - provided that the default for the late time cut is set to infinity or very large times. The calo experts should comment, if they think differently ....

**Comment by ebrianne** [comment]
Hi @rete, this looks good to me. I agree with @gaede that the default time cuts should be put as high as possible. One may be interested in the future to look at late time hit in hadron showers.

File: CaloDigi/Realistic/doc/Readme_ROCTiming.md [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+#### Current ROC parameter values for known technologies (03.2021)
+
+- AHCal:
+  - `fastShaper`: 15 ns
+  - `slowShaper`: 50 ns
+  - `timingResolution`: 700 ps
+- SiWEcal:
+  - `fastShaper`: 90 ns. First estimate
+  - `slowShaper`: 180 ns. First estimate
+  - `timingResolution`: 700 ps. First guess from AHCAL value, as the TDC is similar. Not a real estimate
```
Comment (ebrianne) [inline CaloDigi/Realistic/doc/Readme_ROCTiming.md:52]: Can we get a number of the SiW ECAL prototype? I would think it is in the same ballpark or even better.

Comment (rete) [reply CaloDigi/Realistic/doc/Readme_ROCTiming.md:52]: @airqui ping

**Review by tmadlener — APPROVED** [review]

**Pull request merged**